File: src/problem023/solution023.py
def proper_divisors(n):
    for i in xrange(1, n / 2 + 1):
        if n % i == 0:
            yield i


def main():
    ceiling = 28123
    abundant_nums = [i for i in xrange(2, ceiling) if i < sum(proper_divisors(i))]
    abundant_sums = set((a + b for a in abundant_nums for b in abundant_nums if a + b < ceiling))
    print(sum(set(range(1, ceiling)).difference(abundant_sums)))

if __name__ == '__main__':
    main()

File: src/problem023/solution023.hs
import Data.Set (fromList, toList, difference)

ceil = 28123
main = print . sum . toList $ difference (fromList [1..ceil]) abundantSums
abundantSums = fromList $ [x + y | x <- abundantNums, y <- abundantNums, x + y <= ceil]
abundantNums = filter isAbundant $ [1..ceil]
isAbundant n = n < (sum . filter (\d -> n `mod` d == 0) $ [1..(n `div` 2)])
